### PR TITLE
Adjust the proposal tracking logic

### PR DESCRIPTION
**Pull request by peterwillcn**
Signed-off-by: Xiaobo Tian <peterwillcn@gmail.com>

### CODE REVIEW

File: cr/state/proposalmanager.go [hunk 1]
```diff
@@ -204,12 +201,22 @@ func (p *ProposalManager) transferCRAgreedState(proposal *ProposalState,
 			proposal.Status = CRAgreed
 		})
 	} else {
+		withdrawingBudgets := make(map[uint8]common.Fixed64)
```
Comment (RainFallsSilent) [inline cr/state/proposalmanager.go]: only one imprest payment，no need to use map.

File: cr/state/proposalmanager.go [hunk 2]
```diff
@@ -206,10 +203,19 @@ func (p *ProposalManager) transferCRAgreedState(proposal *ProposalState,
...
+				if b.Type == payload.Imprest {
+					proposal.WithdrawableBudgets[b.Stage] = b.Amount
+				}
+			}
 		}, func() {
 			proposal.Status = CRAgreed
-			proposal.CurrentStage = 0
+			for _, b := range proposal.Proposal.Budgets {
+				if b.Type == payload.Imprest {
+					delete(proposal.WithdrawableBudgets, b.Stage)
```
Comment (Houshoupei84) [inline cr/state/proposalmanager.go]: break will be better

File: cr/state/proposalmanager.go [hunk 3]
```diff
@@ -206,10 +203,19 @@ func (p *ProposalManager) transferCRAgreedState(proposal *ProposalState,
 	} else {
 		p.history.Append(height, func() {
 			proposal.Status = VoterAgreed
-			proposal.CurrentStage = 1
+			for _, b := range proposal.Proposal.Budgets {
+				if b.Type == payload.Imprest {
+					proposal.WithdrawableBudgets[b.Stage] = b.Amount
```
Comment (Houshoupei84) [inline cr/state/proposalmanager.go]: a break will be better

File: core/types/payload/crcproposaltracking.go [hunk 4]
```diff
@@ -35,8 +35,8 @@ const (
 	// charge of the CRC proposal.
 	ProposalLeader CRCProposalTrackingType = 0x04
 
-	// Appropriation indicates that the transaction is used for appropriation.
-	Appropriation CRCProposalTrackingType = 0x05
+	// Finalized indicates that the transaction is used for appropriation.
```
Comment (Houshoupei84) [inline core/types/payload/crcproposaltracking.go]: Comments is not proper

File: core/types/payload/crcproposal.go [hunk 5]
```diff
@@ -152,6 +166,16 @@ func (p *CRCProposal) SerializeUnsigned(w io.Writer, version byte) error {
 	return nil
 }
 
+func (b *Budget) Serialize(w io.Writer) error {
+	if err := common.WriteElement(w, b.Type); err != nil {
+		return errors.New("failed to serialize Type")
+	}
+	if err := common.WriteElement(w, b.Stage); err != nil {
+		return errors.New("failed to serialize Type")
```
Comment (Houshoupei84) [inline core/types/payload/crcproposal.go]: should be  failed to serialize stage

File: blockchain/txvalidator.go [hunk 6]
```diff
@@ -1872,25 +1878,26 @@ func (b *BlockChain) checkCRCProposalTerminatedTracking(
...
 }
 
-func (b *BlockChain) checkCRCProposalAppropriationTracking(
+func (b *BlockChain) checkCRCProposalLeaderTracking(
 	cptPayload *payload.CRCProposalTracking, pState *crstate.ProposalState) error {
 	// Check stage of proposal
-	if cptPayload.Stage != pState.CurrentStage+1 {
-		return errors.New("invalid stage")
+	if cptPayload.Stage != 0 {
+		return errors.New("stage need to be zero")
```
Comment (Houshoupei84) [inline blockchain/txvalidator.go]: stage should not assignment value

File: blockchain/txvalidator.go [hunk 7]
```diff
@@ -1853,8 +1850,17 @@ func (b *BlockChain) checkCRCProposalCommonTracking(
...
 	// Check stage of proposal
-	if cptPayload.Stage != pState.CurrentStage {
-		return errors.New("invalid stage")
+	if _, ok := pState.WithdrawnBudgets[cptPayload.Stage]; ok {
+		return errors.New("invalid budgets with withdrawn budget")
+	}
+
+	for _, budget := range pState.Proposal.Budgets {
+		if cptPayload.Stage == budget.Stage {
+			if budget.Type == payload.Imprest ||
```
Comment (Houshoupei84) [inline blockchain/txvalidator.go]: 1,erros.new string not proper and duplicated

**Pull request merged**